lib/serve/functions.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+var _ = require('lodash');
+var chalk = require('chalk');
+var path = require('path');
+var RSVP = require('rsvp');
+var getProjectId = require('../getProjectId');
+var logger = require('../logger');
+var utils = require('../utils');
+var parseTriggers = require('../parseTriggers');
+var EmulatorController;
+var controller;
+
+function _pollOperation(service, operation) {
+  return new RSVP.Promise(function(resolve, reject) {
+    var poll = function() {
+      service.operations.get({
+        name: operation[0].name
+      }, function(err, _operation) {
+        if (err) {
+          reject(err);
+        } else if (_operation.done) {
+        // TODO: handle deploy failures
+          resolve(_operation.response.value);
+        } else {
+          setTimeout(poll, 500);
+        }
+      });
+    };
+    poll();
+  });
+}
+
+function _startEmulator(options) {
+  try {
+    EmulatorController = require('@google-cloud/functions-emulator/src/cli/controller');
+  } catch (err) {
+    var msg = err;
+    if (process.version !== 'v6.9.1') {
+      msg = 'Please use Node version v6.9.1, you have ' + process.version + '\n';
+    }
+    utils.logWarning('Cannot start functions emulator. ' + msg);
+    return RSVP.reject();
+  }
+  if (process.version !== 'v6.9.1') {
+    utils.logWarning('The functions emulator works best with Node version v6.9.1, you have ' + process.version + '\n');
+  }
+  var functionsDir = path.join(options.config.projectDir, options.config.get('functions.source'));
+  controller = new EmulatorController({verbose: true});
+  // TODO: pipe logs to console
+  logger.info(chalk.bold('Functions Directory:'), options.config.get('functions.source'), '\n');
+
+  return controller.start().then(function() {
+    return controller.clear();
+  }).then(function() {
+    return parseTriggers(getProjectId(options), options.instance, functionsDir);
+  // TODO: display helpful message when there are errors from trigger parsing
+  }).then(function(triggers) {
+    var promises = _.map(triggers, function(trigger) {
+      if (trigger.httpsTrigger) {
+        return controller.deploy(trigger.name, {
```

Comment:
Luckily, the Emulator's codebase has been fairly well decomposed, so instead of starting up the whole Emulator and using the CLI's `Controller` class to communicate with the Emulator, you could `require` the `Supervisor`[1] class and just start that. You'd be bypassing the `Controller` and the Emulated GCP API completely.

[1] The Supervisor is the runtime. It starts and stops the worker processes. It does have a dependency on the Emulator's Model layer though, so using the Supervisor directly wouldn't be _that_ straightforward, but it could be done.